source/emitSchema.ts

```diff
@@ -5,85 +5,112 @@ import type {
+    baseSchema,
+    schema
+  )} {
+  `);
+
   // For each property, add a type to the interface
-  let interfaceContent = convertPropertiesToTypes(
+  emitTypeProperties(
+    typescriptEmitter,
     schema,
     baseSchema?.properties,
     typedContextSchema?.properties
   );
+
   // Entity type and permissions are missing from the source schema, so add them to the interface
-  interfaceContent = addEntityTypeAndPermissions(
-    interfaceName,
-    interfaceContent
-  );
+  emitSpecialProperties(typescriptEmitter, schema);
 
-  //Gets the suffix for the interface, if it extends another schema
-  const interfaceSuffix = getTypeExtensionSuffix(baseSchema, schema);
+  typescriptEmitter.appendCode(`}`);
+}
 
+function getTypeScriptInterfaceNameForInterface(schema: Schema) {
   // Adds a generic to the interface to TypedContext, which is used for subtypes of TypedContext
-  if (interfaceName === "TypedContext") {
-    interfaceName =
-      "TypedContext<K extends TypedContextSubtype = TypedContextSubtype>";
+  if (isSchemaTypedContext(schema)) {
+    return "TypedContext<K extends TypedContextSubtype = TypedContextSubtype>";
   }
-  const conversionErrors = errors;
-  errors = [];
-  const TSInterface = `
-    export interface ${interfaceName} ${interfaceSuffix} {${interfaceContent.join(
-    "; "
-  )}}`;
-  return { TSInterface, conversionErrors };
+
+  return schema.id;
 }
-function getId(schema: Schema) {
-  const id = schema.id;
-  if (!id) {
-    errors.push(`No id defined for schema ${schema.id}`);
-  }
-  return id;
+
+function isSchemaTypedContext(schema: Schema) {
+  return schema.id === "TypedContext";
 }
-function addEntityTypeAndPermissions(
-  interfaceName: string,
-  interfaceContent: string[]
+
+function emitSpecialProperties(
+  typescriptEmitter: TypeScriptEmitter,
+  schema: Schema
 ) {
-  if (interfaceName === "TypedContext") {
-    interfaceContent.push(`__entity_type__?: K`);
+  if (isSchemaTypedContext(schema)) {
+    typescriptEmitter.appendCode(`__entity_type__: K;`);
   } else {
-    interfaceContent.push(`__entity_type__?: "${interfaceName}"`);
+    typescriptEmitter.appendCode(`__entity_type__: "${schema.id}";`);
+  }
+  typescriptEmitter.appendCode(`__permissions?: Record<string, any>;`);
+
+  if (schema.properties && "custom_attributes" in schema.properties) {
+    if (isSchemaTypedContext(schema)) {
+      typescriptEmitter.appendCode(
+        `custom_attributes: Array<TypedContextCustomAttributesMap[K]>;`
+      );
+    } else {
+      typescriptEmitter.appendCode(
+        `custom_attributes: Array<TypedContextCustomAttributesMap["${schema.id}"]>;`
+      );
+    }
   }
-  interfaceContent.push(`__permissions?: Record<string, any>`);
-  return interfaceContent;
 }
-function getTypeExtensionSuffix(baseSchema?: Schema, schema?: Schema) {
+
+function getTypeExtensionSuffix(
+  baseSchema: Schema | undefined,
+  schema: Schema
+) {
+  const omitList = ["__entity_type__", "__persmissions"];
```

Comment:
Should probably not be `__persmissions` ;)

Reply:
Oooh 🤣 nice catch!